src/MultigridMesh.f90

```diff
@@ -0,0 +1,418 @@
+! Futility is a jointly-maintained, open-source project between the University !
+! of Michigan and Oak Ridge National Laboratory.  The copyright and license    !
+! can be found in LICENSE.txt in the head directory of this repository.        !
+!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+!> @brief Module provides 3 types that help define a hierachy of meshes in a
+!>    multigrid context.  It is used to help define the interpolation operators
+!>    for a multigrid method.  Note that the meshes should be thought of as
+!>    points rather than cells.  However, it is generally okay to treat cells
+!>    as points when applying a multigrid method that is not fully geometric.
+!>
+!> For valid reference lists
+!> see @ref MatrixTypes::LinearSolverTypes_Declare_ValidParams
+!> "LinearSolverTypes_Declare_ValidParams".
+!>
+!> Currently supported TPLs include:
+!>  - PETSc (with interfaces to KSP)
+!>
+!> @par Module Dependencies
+!>  - @ref IntrType "IntrType": @copybrief IntrType
+!>  - @ref BLAS "BLAS": @copybrief BLAS
+!>  - @ref Times "Times": @copybrief Times
+!>  - @ref ExceptionHandler "ExceptionHandler": @copybrief ExceptionHandler
+!>  - @ref Allocs "Allocs": @copybrief Allocs
+!>  - @ref ParameterLists "ParameterLists": @copybrief ParameterLists
+!>  - @ref ParallelEnv "ParallelEnv": @copybrief ParallelEnv
+!>  - @ref VectorTypes "VectorTypes": @copybrief VectorTypes
+!>  - @ref MatrixTypes "MatrixTypes": @copybrief MatrixTypes
+!>  - @ref LinearSolverTypes "LinearSolverTypes": @copybrief LinearSolverTypes
+!>
+!> @par EXAMPLES
+!> @code
+!>
+!> @endcode
+!>
+!> @author Ben C. Yee
+!>   @date 09/11/2017
+!>
+!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+
+MODULE MultigridMesh
+  USE IntrType
+  USE ExceptionHandler
+  USE ParameterLists
+  USE VectorTypes
+  USE MatrixTypes
+  USE LinearSolverTypes
+  IMPLICIT NONE
+  PRIVATE
+
+  !PUBLIC :: MultigridMeshElementType
+  PUBLIC :: MultigridMeshType
+  PUBLIC :: MultigridMeshStructureType
+  PUBLIC :: InterpWeightsLevelType
+  PUBLIC :: InterpWeightsStructureType
+
+  TYPE :: MultigridMeshElementType
+    !> Index of the point on the finest grid that this point eventually injects
+    !>  into.  May have another use/definition in cases where the interpolation
+    !>  operator has no injections.  This is not needed for fillInterpMatrices
+    !>  in LinearSolverTypes_Multigrid, it is only here for the user's
+    !>  convenience and using/setting this variable is optional.
+    INTEGER(SIK) :: finestGridIndex=-1
+    !> For degree 0 points, childIndices is a size-1 array containing the
+    !>  **child** mesh index of the point that it injects into.
+    !> For degree >0 points, childIndices is a size >1 array containing the
+    !>  **current** mesh indices of the points from which the value on this point
+    !>  is interpolated.
+    !> For MPACT, ordering should be WSENBT on finest mesh and WESNBT on
```

Comment:
I think this comment would probably be more appropriate somewhere in MPACT where this is set up, since a Futility use may not know or care anything about MPACT.

Also, was this numbering scheme chosen for a particular purpose?  In MPACT, we enumerate WNESTB.  May not be a big deal, but make sure you're clear wherever this gets used in MPACT or people will break it eventually.

Reply:
I'll move this comment to MPACT.  It was more of a note to myself.  In general, it is arbitrary as long as you are consistent.  From the multigrid linear solver's perspective, it is irrelevant even if you are inconsistent.  Maintaining consistency is mostly to help you avoid errors when filling out multigridmesh levels.

Regarding the convention MPACT, my "positive" x and y directions are EAST and NORTH; I wanted the "negative" directions listed first where possible.  This is really only relevant for CMFD_Std::fillMultigridMeshes.

Reply:
Ok.  The reason for the convention was more of a curiosity on my part.  But yeah, moving the comment to MPACT so others know what the convention is and why it was chosen is good.  No need to actually change that convention.